### PR TITLE
FifoRecorder: Use Texture Cache to record efb copies correctly.

### PR DESCRIPTION
**Pull request by phire**

### CODE REVIEW

File: Source/Core/Core/FifoPlayer/FifoRecordAnalyzer.cpp [hunk 1]
```diff
@@ -227,78 +228,3 @@ void FifoRecordAnalyzer::WriteVertexArray(int arrayIndex, u8 *vertexData, int ve
 
 	FifoRecorder::GetInstance().WriteMemory(arrayStart, arraySize, MemoryUpdate::VERTEX_STREAM);
 }
-
-void FifoRecordAnalyzer::ProcessTexMaps()
```
Comment (degasus) [inline Source/Core/Core/FifoPlayer/FifoRecordAnalyzer.cpp]: Do you think we should merge this PR as it is, or also wait for the next commits which inlines the functions above?

Comment (phire) [inline Source/Core/Core/FifoPlayer/FifoRecordAnalyzer.cpp]: They don't interact badly with each other, so it will be fine to merge this as is.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 2]
```diff
@@ -631,6 +635,14 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
...
 			memcpy(temp, l.data, l.data_size);
 			entry->Load(l.width, l.height, l.width, level);
+
+			// We still need to calculate the original size of the mips
+			const u32 mip_width = CalculateLevelSize(width, level);
+			const u32 mip_height = CalculateLevelSize(height, level);
+			const u32 expanded_mip_width = ROUND_UP(mip_width, bsw);
+			const u32 expanded_mip_height = ROUND_UP(mip_height, bsh);
+
+			mips_size += TexDecoder_GetTextureSizeInBytes(expanded_mip_width, expanded_mip_height, texformat);
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I think it's better to move this loop out of the hires code and into an if(g_bRecordFifoData) block.

File: Source/Core/Core/FifoPlayer/FifoRecorder.cpp [hunk 3]
```diff
@@ -102,6 +102,25 @@ void FifoRecorder::WriteGPCommand(u8 *data, u32 size)
 	m_SkipNextData = m_SkipFutureData;
 }
 
+void FifoRecorder::ShadowMemory(u32 address, u32 size)
+{
+	u8 *curData;
+	u8 *newData;
```
Comment (lioncash) [inline Source/Core/Core/FifoPlayer/FifoRecorder.cpp]: asterisks against type name

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 2 above)
Comment (phire) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I think I'll move it into a function and calculate it next to texture_size. mips_size will be useful for other things later.

Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: I guess with level=0, this function is the same as texture_size.

Comment (phire) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: No, it's only the size of the additional mips. If there is only one level this will be zero.

File: Source/Core/VideoCommon/VideoCommon.h [hunk 4]
```diff
@@ -12,6 +12,9 @@
 #include "Common/MathUtil.h"
 #include "VideoCommon/VideoBackendBase.h"
 
+// Global flag to signal if FifoRecorder is active.
+extern bool g_bRecordFifoData;
```
Comment (degasus) [inline Source/Core/VideoCommon/VideoCommon.h]: FifoRecorder.h?

Comment (phire) [inline Source/Core/VideoCommon/VideoCommon.h]: The flag belongs to video common. Toggled in RenderBase.cpp

**Comment by phire** [comment]
Ok, fixed up those issues.

File: Source/Core/Core/FifoPlayer/FifoRecordAnalyzer.cpp (hunk 1 above)
Comment (phire) [inline Source/Core/Core/FifoPlayer/FifoRecordAnalyzer.cpp]: For future reference, the "above functions" (ProcessPreloadTexture and ProcessLoadTlut1) were in fact moved to VideoCommon in later commits.

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 5]
```diff
@@ -687,6 +700,10 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
 		}
 	}
 
+	// If we are recording a FifoLog, keep track of what memory we read.
+	if (g_bRecordFifoData && !from_tmem)
+		FifoRecorder::GetInstance().WriteMemory(address, texture_size + additional_mips_size, MemoryUpdate::TEXTURE_MAP);
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: This could be moved to the top of this function. It will slow down while the fifo is recorded, but so it won't catch hashing issues,

File: Source/Core/VideoCommon/TextureCacheBase.cpp [hunk 6]
```diff
@@ -678,6 +690,7 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
 				: src_data;
 			const u8* tlut = &texMem[tlutaddr];
 			TexDecoder_Decode(temp, mip_src_data, expanded_mip_width, expanded_mip_height, texformat, tlut, (TlutFormat)tlutfmt);
+
```
Comment (degasus) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: unrelated whitespace change.

**Comment by degasus** [comment]
Else LGTM

File: Source/Core/VideoCommon/TextureCacheBase.cpp (hunk 5 above)
Comment (phire) [inline Source/Core/VideoCommon/TextureCacheBase.cpp]: Hmm, that is a good point.

File: Source/Core/Core/FifoPlayer/FifoRecordAnalyzer.cpp (hunk 1 above)
Comment (degasus) [inline Source/Core/Core/FifoPlayer/FifoRecordAnalyzer.cpp]: "ProcessLoadIndexedXf" and "ProcessVertexArrays" are still missing. To be honest, they are the easier ones :D

Comment (phire) [inline Source/Core/Core/FifoPlayer/FifoRecordAnalyzer.cpp]: This PR is texture cache only.

**Comment by phire** [comment]
Slight cleanup to merge `ShadowMemory` and `WriteMemory` into `UseMemory` And eliminate duplicated code. 

Also Texture Cache now records memory updates before checking the hash, so the correct thing will get recorded even if the texture cache hash algorithm is wrong.
